### PR TITLE
chore: Use cypress for storybook visual tests

### LINKED ISSUE #640: Investigate Cypress for visual testing

### PR DESCRIPTION
**Pull request by mayank99**
This PR introduces a new `apps/storybook` workspace dedicated for Storybook and Cypress visual testing.

### Storybook

The `storybook` app is initialized with a minimal `vite` app and the storybook is built using `@storybook/builder-vite` instead of webpack.

After this, I copied over our existing setup from `packages/itwinui-react` and deleted it from there.

There are a bunch of changes in story files which mainly fall under these 4 categories:
  - importing components from `@itwin/itwinui-react`
  - removing creevey stuff
  - fixing random typescript errors
  - replacing `useEffect` from storybook (doesn't work in vite) with the one from react

I've also added the Open Sans font files in the public directory and the corresponding `@font-face` declarations in `preview-head`.

Storybook can be built using `yarn build` script and local server with HMR can be started using `yarn dev`. This ensures that these commands are included in the turborepo root runs.

### Cypress

Initially I tried using cypress [_component testing_](https://docs.cypress.io/guides/component-testing/introduction) by composing stories using `@storybook/testing-react` and mounting them using `@cypress/react`. This was giving me a bunch of errors which I couldn't fix so I decided to abandon this and go with the traditional "integration testing" route by using `cy.visit()` on story iframes. The actual visual screenshot testing is done using [`cypress-image-diff-js`](https://github.com/uktrade/cypress-image-diff).

Each story has a corresponding `.test.ts` file which explicitly lists the story names. Any stories that need to be skipped can simply be excluded from this list. Any interactions previously performed in creevey/selenium are now performed in this file as well.

Test images are currently generated using Electron 94 browser. This seems to be the latest version included in Cypress. 

### Docker

I reused the Dockerfile from [`iTwinUI-layouts`](https://github.com/iTwin/iTwinUI-layouts/) and updated Cypress version, so if we decide to host it on Dockerhub, we should use the same image in both places. Currently it needs to be built (once) locally using `yarn build-docker` which takes around 10-12 mins on M1 Mac, and once on CI using [`build-push-action`](https://github.com/docker/build-push-action) which takes 15-20 mins, after which it is cached.

The `run-tests.js` file loads this docker image, and then `entrypoint.sh` uses [`serve`](https://github.com/vercel/serve) to start an http server on the built `storybook-static` folder, before running `cypress run`. The tests currently take ~6 mins to finish and requires `build` to be run first, which takes ~2 mins. This could be improved by [parallelization](https://docs.cypress.io/guides/guides/parallelization) but that might require publishing the Docker image.

### Misc / unrelated changes

I moved the storybook eslint plugin out of our shared `configs` because it's only used in one place. So I had to also add `eslint-config-prettier` as a peer dep of `configs` because I was getting [random errors](https://user-images.githubusercontent.com/9084735/166827309-4715e86d-075b-4d5f-b0d3-20634bcee116.png).

Also had to add a check for `NaN` in `useOverflow` because there was some division-by-zero happening.

### CODE REVIEW

File: .github/workflows/publish-storybook.yml [hunk 1]
```diff
@@ -20,10 +20,10 @@ jobs:
           cache: yarn
 
       - run: yarn install --frozen-lockfile
-      - run: yarn workspace @itwin/itwinui-react build-storybook
+      - run: yarn build
```
Comment (veekeys) [inline .github/workflows/publish-storybook.yml:23]: sure we need to build everything now?

**Review by veekeys — COMMENTED** [review]
I have reviewed code, but havent run it locally yet.
Together with the comments, we need to update fully our readmes whenever its applicable.
Havent reviewed deeply scripts and test files code, but went through images.

File: package.json [hunk 2]
```diff
@@ -14,9 +14,25 @@
     "test": "turbo run test",
     "lint-staged": "turbo run lint-staged"
   },
+  "dependencies": {},
   "devDependencies": {
-    "turbo": "~1.2.4",
-    "webpack": "5"
+    "turbo": "~1.2.4"
```
[the remark this answers is not in the document]
Comment (mayank99) [reply package.json:19]: yes

File: packages/iTwinUI-react/stories/intro/Overview.stories.tsx [hunk 3]
```diff
@@ -1,115 +0,0 @@
-/*---------------------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (mayank99) [reply packages/iTwinUI-react/stories/intro/Overview.stories.tsx:1]: i plan to add it back. That's why it says "pending" in the PR description

File: packages/iTwinUI-react/src/core/utils/hooks/useOverflow.tsx [hunk 4]
```diff
@@ -85,8 +85,11 @@ export const useOverflow = <T extends HTMLElement>(
       );
       const avgItemSize = childrenSize / visibleCount;
       const visibleItems = Math.floor(availableSize / avgItemSize);
-      // Doubling the visible items to overflow the container. Just to be safe.
-      setVisibleCount(Math.min(items.length, visibleItems * 2));
+
+      if (!isNaN(visibleItems)) {
```
[the remark this answers is not in the document]
Comment (mayank99) [reply packages/iTwinUI-react/src/core/utils/hooks/useOverflow.tsx:89]: dividing by zero was resulting in `NaN`

File: packages/configs/package.json [hunk 5]
```diff
@@ -14,8 +14,11 @@
     "eslint-config-prettier": "^8.3.0",
     "eslint-plugin-react": "^7.29.0",
     "eslint-plugin-react-hooks": "^4.2.0",
-    "eslint-plugin-storybook": "^0.5.7",
     "prettier": "2.2.1"
   },
+  "peerDependencies": {
```
[the remark this answers is not in the document]
Comment (mayank99) [reply packages/configs/package.json:19]: I only added it here to prevent any misuse, as it will prompt to install eslint when you try using it alone

File: apps/storybook/src/ColorPicker.test.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+describe('ColorPicker', () => {
+  const storyPath = 'Core/ColorPicker';
+  const tests = ['Basic', 'Advanced', 'With Alpha'];
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/ColorPicker.test.ts:3]: i think it's focused and selected

File: .vscode/settings.json [hunk 7]
```diff
@@ -11,5 +11,5 @@
     "source.fixAll.eslint": true
   },
   "files.eol": "\n",
-  "typescript.tsdk": "node_modules\\typescript\\lib"
+  "typescript.tsdk": "node_modules/typescript/lib"
```
[the remark this answers is not in the document]
Comment (mayank99) [reply .vscode/settings.json:14]: backslash doesn't work in macos

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -63,6 +62,25 @@ jobs:
...
+      - uses: docker/setup-buildx-action@v1
+
+      - uses: actions/cache@v3
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-${{ hashFiles('apps/storybook/scripts/docker/Dockerfile') }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-
+
+      - uses: docker/build-push-action@v2
```
[the remark this answers is not in the document]
Comment (mayank99) [reply .github/workflows/build.yml:75]: well yeah, CI is passing

File: apps/storybook/src/Toasts.stories.tsx [hunk 9]
```diff
@@ -4,22 +4,20 @@
...
-import { CreeveyMeta } from 'creevey';
-import { ToasterSettings } from '../../src/core/Toast/Toaster';
+import { Button, toaster, ProgressRadial } from '@itwin/itwinui-react';
+
+import { ToastProps } from '@itwin/itwinui-react/esm/core/Toast/Toast';
+import { ToasterSettings } from '@itwin/itwinui-react/esm/core/Toast/Toaster';
 
 export default {
-  subcomponents: { Toast },
   argTypes: {
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Toasts.stories.tsx:13]: it wasn't showing any props when clicking on Toast, and it also feels duplicated

File: apps/storybook/src/Toasts.stories.tsx [hunk 10]
```diff
@@ -4,22 +4,20 @@
  *--------------------------------------------------------------------------------------------*/
 import React from 'react';
 import { Story, Meta } from '@storybook/react';
-import { Button, toaster, ProgressRadial } from '../../src/core';
-import { Toast, ToastProps } from '../../src/core/Toast/Toast';
-import { CreeveyMeta } from 'creevey';
-import { ToasterSettings } from '../../src/core/Toast/Toaster';
+import { Button, toaster, ProgressRadial } from '@itwin/itwinui-react';
+
+import { ToastProps } from '@itwin/itwinui-react/esm/core/Toast/Toast';
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Toasts.stories.tsx:9]: ?

File: apps/storybook/src/Tile.stories.tsx [hunk 11]
```diff
@@ -174,8 +178,12 @@ export const Actionable: Story<TileProps> = (props) => {
       metadata={metadata}
       badge={badge}
       thumbnail={thumbnail}
+      buttons={buttons}
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Tile.stories.tsx]: Was a bad merge. Fixed

File: apps/storybook/src/Table.stories.tsx [hunk 12]
```diff
@@ -1740,9 +1632,9 @@ WithPaginator.decorators = [
 
 WithPaginator.argTypes = {
   data: { control: { disable: true } },
-  parameters: {
-    docs: { source: { excludeDecorators: true } },
-  },
+};
+WithPaginator.parameters = {
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Table.stories.tsx:1636]: It was incorrect syntax

File: apps/storybook/src/InputGroup.test.ts [hunk 13]
```diff
@@ -0,0 +1,15 @@
+describe('InputGroup', () => {
+  const storyPath = 'Input/InputGroup';
+  const tests = ['Checkbox Group', 'Radio Group', 'Toggle Group'];
+
+  tests.forEach((testName) => {
+    const id = `${storyPath
+      .replace('/', '-')
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/InputGroup.test.ts:7]: It's just a single string replace, is it worth moving into a helper? Then I would have to import it in every file. Unless you want me to create a custom cypress command or something

File: apps/storybook/scripts/build-docker.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+const os = require('os');
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/scripts/build-docker.js:5]: copied and modified a little bit. layouts copied it from itwinui-css.

File: apps/storybook/src/Textarea.test.ts [hunk 15]
```diff
@@ -0,0 +1,15 @@
+describe('Textarea', () => {
+  const storyPath = 'Input/Textarea';
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Textarea.test.ts:2]: i prefer flat list because the folders were very random and inconsistent

File: apps/storybook/src/Text.test.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+describe('Text', () => {
+  const storyPath = 'Typography/Text';
+  const tests = ['Basic', 'Polymorphic', 'Skeleton'];
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Text.test.ts:3]: no, it has never failed for me. animations are stopped before taking the screenshot

File: apps/storybook/src/SplitButton.test.ts [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+  tests.forEach((testName) => {
+    const id = `${storyPath
+      .replace('/', '-')
+      .toLowerCase()}--${testName.replaceAll(' ', '-').toLowerCase()}`;
+
+    it(testName, () => {
+      cy.visit('iframe', { qs: { id } });
+      cy.compareSnapshot(`${testName} (Closed)`);
+      cy.get('.iui-button').last().click();
+      cy.compareSnapshot(`${testName} (Open)`);
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/SplitButton.test.ts:14]: i couldn't get it to take the current width, and also couldn't reproduce this case in real browser

File: apps/storybook/src/DatePicker.test.ts [hunk 18]
```diff
@@ -0,0 +1,17 @@
+describe('DatePicker', () => {
+  const storyPath = 'Core/DatePicker';
+  const tests = ['Basic', 'Localized', 'With Time', 'With Year'];
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/DatePicker.test.ts:3]: font not present in docker image?

File: .github/workflows/publish-storybook.yml (hunk 1 above)
Comment (mayank99) [reply .github/workflows/publish-storybook.yml]: Added `--filter=storybook`

File: apps/storybook/src/Text.stories.tsx [hunk 19]
```diff
@@ -46,7 +43,7 @@ Polymorphic.args = {
   children: "I'm a headline text rendered as an h4 element!",
   variant: 'headline',
   as: 'h4',
-};
+} as TextProps<'h4'>; // need this cast because TS is complaining about `as`
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Text.stories.tsx:46]: it works in the component, bot sure what happened here

File: apps/storybook/src/Select.stories.tsx [hunk 20]
```diff
@@ -195,13 +164,13 @@ Disabled.args = {
...
 
 Disabled.parameters = {
-  creevey: {
-    tests: {
-      async open() {
-        await this.expect(await this.takeScreenshot()).to.matchImage('open');
-      },
-    },
-  },
+  // creevey: {
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Select.stories.tsx]: removed

File: apps/storybook/src/LabeledTextarea.stories.tsx [hunk 21]
```diff
@@ -28,22 +27,22 @@ export default {
 export const Basic: Story<LabeledTextareaProps> = (args) => {
   return (
     <LabeledTextarea
-      label='Textarea Label'
       message='Display Message'
       placeholder='This is a textarea'
       {...args}
+      label='Textarea Label'
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/LabeledTextarea.stories.tsx:33]: I was getting this TS error:

```
'label' is specified more than once, so this usage will be overwritten.
```

File: apps/storybook/src/ProgressLinear.test.ts [hunk 22]
```diff
@@ -0,0 +1,23 @@
+describe('ProgressLinear', () => {
+  const storyPath = 'ProgressIndicators/ProgressLinear';
+  const tests = [
+    'Determinate',
+    'Determinate Animated',
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/ProgressLinear.test.ts:5]: Removed

File: apps/storybook/src/Table.stories.tsx [hunk 23]
```diff
@@ -34,11 +34,11 @@ import {
   Anchor,
   SelectionColumn,
   ExpanderColumn,
-} from '../../src/core';
+} from '@itwin/itwinui-react';
 import { Story, Meta } from '@storybook/react';
 import { useMemo, useState } from '@storybook/addons';
 import { action } from '@storybook/addon-actions';
-import { CreeveyMeta, CreeveyStoryParams } from 'creevey';
+
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Table.stories.tsx]: Removed all empty lines

File: apps/storybook/src/Select.test.ts [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+  ];
+
+  tests.forEach((testName) => {
+    const id = `${storyPath
+      .replace('/', '-')
+      .toLowerCase()}--${testName.replaceAll(' ', '-').toLowerCase()}`;
+
+    it(testName, () => {
+      cy.visit('iframe', { qs: { id } });
+      cy.compareSnapshot(`${testName} (Closed)`);
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Select.test.ts:35]: Should be fixed now

File: apps/storybook/src/Select.test.ts [hunk 25]
```diff
@@ -0,0 +1,27 @@
+describe('Select', () => {
+  const storyPath = 'Input/Select';
+  const tests = [
+    'Basic',
+    'Custom',
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Select.test.ts:5]: Added `With Value` state for `Custom` case

File: apps/storybook/src/Wizard.test.ts [hunk 26]
```diff
@@ -0,0 +1,21 @@
+describe('Wizard', () => {
+  const storyPath = 'Core/Wizard';
+  const tests = [
+    'Basic',
+    'Long',
+    'Localized Long',
+    'With Tooltips',
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Wizard.test.ts:7]: Fixed

File: apps/storybook/src/UserIconGroup.test.ts [hunk 27]
```diff
@@ -0,0 +1,21 @@
+describe('UserIconGroup', () => {
+  const storyPath = 'Core/UserIconGroup';
+  const tests = [
+    'Basic',
+    'Animated',
+    'Many Icons',
+    'Non Stacked',
+    'With Tooltip',
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/UserIconGroup.test.ts:8]: Fixed

File: apps/storybook/src/InformationPanel.test.ts [hunk 28]
```diff
@@ -0,0 +1,16 @@
+describe('InformationPanel', () => {
+  const storyPath = 'Core/InformationPanel';
+  const tests = ['Basic', 'Custom Actions', 'Custom Width', 'Horizontal'];
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/InformationPanel.test.ts:3]: Fixed now.

File: apps/storybook/src/LabeledTextarea.stories.tsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (veekeys) [reply apps/storybook/src/LabeledTextarea.stories.tsx:33]: ah... I remember seeing it somewhere...
@bentleyvk maybe you know this one?

File: apps/storybook/src/InputGroup.test.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (veekeys) [reply apps/storybook/src/InputGroup.test.ts:7]: I'd have said if cypress has some middleware we can add so we dont even think about it..
Need to also update `createCompnent` script (can add an item for future)

File: .github/workflows/publish-storybook.yml (hunk 1 above)
Comment (veekeys) [reply .github/workflows/publish-storybook.yml]: I remember we had some issues in CI where we missed the `build` command.
Dont remmeber details now, but if you can check that this is not influenced right now, would be good

File: apps/storybook/scripts/build-docker.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (veekeys) [reply apps/storybook/scripts/build-docker.js:5]: Anything particular you think is worth reviewing or if it works should be fine?

File: apps/storybook/src/Tile.test.ts [hunk 29]
```diff
@@ -0,0 +1,22 @@
+describe('Tile', () => {
+  const storyPath = 'Core/Tile';
+  const tests = [
+    'Basic',
```
[the remark this answers is not in the document]
Comment (veekeys) [reply apps/storybook/src/Tile.test.ts:4]: just thought if we have lets say a pixel difference (in spacing or smth), would it fail the test or it might be too small diff as the image taken is much bigger than the component.

File: apps/storybook/src/Textarea.test.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (veekeys) [reply apps/storybook/src/Textarea.test.ts:2]: hm..
I guess we can move those later on in something not random

File: apps/storybook/src/DatePicker.test.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (veekeys) [reply apps/storybook/src/DatePicker.test.ts:3]: interesting...

File: apps/storybook/src/Table.stories.tsx (hunk 23 above)
[the remark this answers is not in the document]
Comment (veekeys) [reply apps/storybook/src/Table.stories.tsx]: Hehe, nice try 
https://github.com/iTwin/iTwinUI-react/pull/660#discussion_r869277241

File: apps/storybook/src/Modal.test.ts [hunk 30]
```diff
@@ -0,0 +1,21 @@
+describe('Modal', () => {
+  const storyPath = 'Core/Modal';
+  const tests = [
+    'Basic',
+    'Full Page Modal',
+    'Non Dismissible Modal',
```
[the remark this answers is not in the document]
Comment (veekeys) [reply apps/storybook/src/Modal.test.ts]: I have put into our `ideas`

File: apps/storybook/src/Table.test.ts [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+        case 'Expandable Subrows': {
+          cy.get('.iui-row-expander').first().click();
+          break;
+        }
+        case 'Editable': {
+          cy.get('[contenteditable]').first().click().type('Test');
+          break;
+        }
+        case 'Filters': {
+          cy.get('.iui-filter-button').first().click({ force: true }); // force because the button is hidden
```
[the remark this answers is not in the document]
Comment (veekeys) [reply apps/storybook/src/Table.test.ts:55]: Added to `ideas`

File: apps/storybook/src/Table.test.ts [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+    'Expandable Subrows',
+    'Filters',
+    'Full',
+    'Horizontal Scroll',
+    'Loading',
+    'No Data',
+    'Resizable Columns',
+    'Selectable Multi',
+    'Selectable Single',
+    'Sortable',
```
[the remark this answers is not in the document]
Comment (veekeys) [reply apps/storybook/src/Table.test.ts:22]: Added to `ideas`

File: apps/storybook/src/Toasts.stories.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (veekeys) [reply apps/storybook/src/Toasts.stories.tsx:8]: esm import, but it is not shown in code anywhere, so let it be...

File: apps/storybook/src/Toasts.stories.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (veekeys) [reply apps/storybook/src/Toasts.stories.tsx:12]: Wasnt showing where?
image

File: apps/storybook/package.json [hunk 33]
```diff
@@ -0,0 +1,51 @@
+{
+  "name": "storybook",
```
Comment (veekeys) [inline apps/storybook/package.json:2]: I had a flood of these warning and it took ages to load up storybook...

File: apps/storybook/src/LabeledTextarea.stories.tsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (bentleyvk) [reply apps/storybook/src/LabeledTextarea.stories.tsx:33]: It was in Table stories. We made our Table props in stories with `Partial<>`

File: apps/storybook/package.json (hunk 33 above)
Comment (veekeys) [reply apps/storybook/package.json:2]: Also, built docker, run `yarn test`, but all tests failed.
Am I missing something?

File: packages/iTwinUI-react/src/core/utils/hooks/useOverflow.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (mayank99) [reply packages/iTwinUI-react/src/core/utils/hooks/useOverflow.tsx:89]: I added this because visual tests were failing. Try removing this line and run storybook, it will show only one breadcrumb

File: apps/storybook/package.json (hunk 33 above)
Comment (mayank99) [reply apps/storybook/package.json:2]: it needs storybook `build` first. where did you run `yarn test`? i think build should automatically happen if you run it from root using `yarn test --filter=storybook`

File: apps/storybook/src/Table.stories.tsx (hunk 23 above)
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Table.stories.tsx]: ok done

File: .github/workflows/publish-storybook.yml (hunk 1 above)
Comment (veekeys) [reply .github/workflows/publish-storybook.yml]: oh.. build included copying at first (readme), but then it failed, cause it did not include changelog or something :D so we changed to copy files.. 
I dont remember truly :D

File: .github/workflows/build.yml [hunk 34]
```diff
@@ -71,6 +89,5 @@ jobs:
 
       - run: yarn install --frozen-lockfile
 
-      # `&` symbol is needed as it prevents Storybook from running endlessly
-      - run: yarn dev &
-      - run: yarn workspace @itwin/itwinui-react creevey
+      - run: yarn build --filter=storybook
```
Comment (bentleyvk) [inline .github/workflows/build.yml:92]: What is the difference between running commands like this versus `yarn workspace...`?

File: apps/storybook/index.html [hunk 35]
```diff
@@ -0,0 +1,20 @@
+<!DOCTYPE html>
```
Comment (bentleyvk) [inline apps/storybook/index.html:1]: Is this file needed?

File: apps/storybook/package.json [hunk 36]
```diff
@@ -0,0 +1,51 @@
...
+    "eslint-plugin-storybook": "^0.5.7",
+    "react": "^17.0.0",
+    "react-dom": "^17.0.0",
+    "serve": "^13.0.2",
+    "storybook-dark-mode": "^1.1.0",
+    "typescript": "^4.6.3",
+    "vite": "^2.8.0"
+  },
+  "scripts": {
+    "dev:vite": "vite",
```
Comment (bentleyvk) [inline apps/storybook/package.json:33]: Why `:vite` postfix?

File: apps/storybook/src/Alert.test.ts [hunk 37]
```diff
@@ -0,0 +1,22 @@
+describe('Alert', () => {
```
Comment (bentleyvk) [inline apps/storybook/src/Alert.test.ts:1]: No copyrights header

File: apps/storybook/package.json (hunk 36 above)
Comment (mayank99) [reply apps/storybook/package.json:33]: because we dont really care about vite commands. I could remove them completely but left them just in case

File: apps/storybook/src/Badge.test.ts [hunk 38]
```diff
@@ -0,0 +1,15 @@
+describe('Badge', () => {
+  const storyPath = 'Core/Badge';
+  const tests = ['Basic', 'Long Label', 'Statuses'];
+
+  tests.forEach((testName) => {
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Badge.test.ts:5]: see https://github.com/iTwin/iTwinUI-react/pull/660#discussion_r869272468

do you have a more specific suggestion of the best way to do this? i want to avoid having to import the helper function in each file

File: apps/storybook/src/Alert.test.ts [hunk 39]
```diff
@@ -0,0 +1,22 @@
...
+  const storyPath = 'Core/Alert';
+  const tests = [
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Alert.test.ts:3]: I was not able to make it work, maybe you know a better way. But even then, we would need to manually exclude the stories we don't want to test.

But I also didn't spend too much time on it because I wanted to avoid adding any storybook-specific logic (right now it's just urls) and we might move off storybook in the future.

File: apps/storybook/src/Badge.test.ts (hunk 38 above)
[the remark this answers is not in the document]
Comment (bentleyvk) [reply apps/storybook/src/Badge.test.ts:5]: Why it is bad to import some helper versus having to copy paste everything and if we want to adjust something there, the good luck?

File: apps/storybook/package.json (hunk 36 above)
Comment (bentleyvk) [reply apps/storybook/package.json:33]: Do we even need them and do they work?

File: packages/iTwinUI-react/src/core/utils/hooks/useOverflow.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (veekeys) [reply packages/iTwinUI-react/src/core/utils/hooks/useOverflow.tsx:89]: we were running it before, but not sure we got it?

**Comment by mayank99** [comment]
Apparently I am using cypress 9.6.0 in the Dockerfile, whereas iTwinUI-layouts is using 9.5.0, so image will need to be rebuilt. We should align the versions in the future.

**Comment by bentleyvk** [comment]
I got this after each test case run:
image

File: apps/storybook/src/Alert.test.ts [hunk 40]
```diff
@@ -0,0 +1,22 @@
...
+    'Sticky',
+  ];
+
+  tests.forEach((testName) => {
+    const id = `${storyPath
+      .replace('/', '-')
+      .toLowerCase()}--${testName.replaceAll(' ', '-').toLowerCase()}`;
+
+    it(testName, () => {
+      cy.visit('iframe', { qs: { id } });
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Alert.test.ts:18]: `serve` removes `.html` from urls, so this shouldn't be an issue. https://github.com/vercel/serve-handler#cleanurls-booleanarray

Did you build storybook before running tests?

[the remark this answers is not in the document]
Comment (bentleyvk) [reply apps/storybook/src/Alert.test.ts:18]: I built, same.

File: apps/storybook/src/Alert.test.ts (hunk 39 above)
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Alert.test.ts:3]: I had already tried that but couldn't get it to work. It tries to import and process the entire file, which doesn't work because it has css imports.

Oops...we found an error preparing this test file: src/Alert.test.ts
The error was:
Error: Webpack Compilation Error
/e2e/node_modules/@itwin/itwinui-css/css/table.css 5:0
Module parse failed: Unexpected character '@' (5:0)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file.

File: apps/storybook/src/Badge.test.ts (hunk 38 above)
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Badge.test.ts:5]: Added `storyId` helper function into the `Cypress` object

File: apps/storybook/package.json (hunk 36 above)
Comment (mayank99) [reply apps/storybook/package.json]: They work but the vite app is just hello world. Removed

File: apps/storybook/index.html (hunk 35 above)
Comment (mayank99) [reply apps/storybook/index.html]: Tests apparently work without it, so removed

File: apps/storybook/src/InputGroup.test.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/InputGroup.test.ts]: Added `storyId` helper function into the `Cypress` object

File: apps/storybook/src/Tile.test.ts (hunk 29 above)
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/storybook/src/Tile.test.ts:4]: I tested with a small pixel difference: Removed the `.` from "More Info." in `Alert`.

Test failed successfully 👍
image

File: apps/storybook/package.json (hunk 33 above)
Comment (mayank99) [reply apps/storybook/package.json:2]: Running `yarn test --filter=storybook` from root is all that's needed after I improved the turbo pipeline in [`5aa3432`](https://github.com/iTwin/iTwinUI-react/pull/660/commits/5aa343282eeb6f7eb44a3feaf883c4318345c1a8).

For running from `apps/storybook`, you still need to manually build storybook first. If you want I could add an alias here, something like:
```
"yarn test:full": "yarn build && yarn test"
```

File: apps/storybook/src/Toasts.stories.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (veekeys) [reply apps/storybook/src/Toasts.stories.tsx:12]: https://itwin.github.io/iTwinUI-react/?path=/docs/core-toasts--anchor-to-button

Here its shown to you as empty???

Minor, I guess if there are issues, can be removed..

File: apps/storybook/package.json (hunk 33 above)
Comment (veekeys) [reply apps/storybook/package.json:2]: We must document all of that..

**Review by veekeys — APPROVED** [review]
Great job!
I left some comments open, but nothing that critical. Most bothering I guess is this
https://github.com/iTwin/iTwinUI-react/pull/660#discussion_r869300977

But I guess for future.
Next we really need to document it all and continue improving the run time.

**Comment by adamhock** [comment]
> > I got this after each test case run: image
> 
> @bentleyvk @adamhock Can you pull the latest changes and see if you still get that error? I think cypress just needs `cypress/tsconfig.json`.

My tests still get the same error after pulling these changes

**Comment by bentleyvk** [comment]
> I got this after each test case run: image

It seems that `extends` in `apps/storybook/tsconfig.json` causes the issue because when I removed and hardcoded rules from the config tsconfig, tests are running normally.

**Comment by mayank99** [comment]
> It seems that `extends` in `apps/storybook/tsconfig.json` causes the issue because when I removed and hardcoded rules from the config tsconfig, tests are running normally.

Ok good, so I'm duplicating the rules now. Can you please verify one last time that it works? @bentleyvk @adamhock

**Comment by bentleyvk** [comment]
> > It seems that `extends` in `apps/storybook/tsconfig.json` causes the issue because when I removed and hardcoded rules from the config tsconfig, tests are running normally.
> 
> Ok good, so I'm duplicating the rules now. Can you please verify one last time that it works? @bentleyvk @adamhock

It is running now for me.

**Review by bentleyvk — APPROVED** [review]

**Review by adamhock — APPROVED** [review]
